InnerEye/ML/SSL/utils.py

```
@@ -0,0 +1,123 @@
+
+import logging
+from enum import Enum
+from pathlib import Path
+from typing import Any, Optional
+
+import torch
+from yacs.config import CfgNode
+
+from InnerEye.ML.SSL import ssl_augmentation_config
+from InnerEye.ML.lightning_container import LightningModuleWithOptimizer
+
+
+class SSLDataModuleType(Enum):
+    ENCODER = 'encoder'
+    LINEAR_HEAD = 'linear_head'
+
+
+class SSLTrainingType(Enum):
+    SimCLR = "SimCLR"
+    BYOL = "BYOL"
+
+
+def load_ssl_augmentation_config(config_path: Path) -> CfgNode:
+    """
+    Loads configs required for self supervised learning. Does not setup cudann as this is being
+    taken care of by lightning.
+    """
+    config = ssl_augmentation_config.get_default_model_config()
+    config.merge_from_file(config_path)
+    config.freeze()
+    return config
+
+
+def create_ssl_encoder(encoder_name: str, use_7x7_first_conv_in_resnet: bool = True) -> torch.nn.Module:
+    """
+    Creates SSL encoder.
+    :param encoder_name: available choices: resnet18, resnet50, resnet101 and densenet121.
+    :param use_7x7_first_conv_in_resnet: If True, use a 7x7 kernel (default) in the first layer of resnet. 
+    If False, replace first layer by a 3x3 kernel. This is required for small CIFAR 32x32 images to not shrink them.
+    """
+    from pl_bolts.models.self_supervised.resnets import resnet18, resnet50, resnet101
+    from InnerEye.ML.SSL.encoders import DenseNet121Encoder
+    if encoder_name == 'resnet18':
+        encoder = resnet18(return_all_feature_maps=False, first_conv=use_7x7_first_conv_in_resnet)
+    elif encoder_name == 'resnet50':
+        encoder = resnet50(return_all_feature_maps=False, first_conv=use_7x7_first_conv_in_resnet)
+    elif encoder_name == 'resnet101':
+        encoder = resnet101(return_all_feature_maps=False, first_conv=use_7x7_first_conv_in_resnet)
+    elif encoder_name == 'densenet121':
+        encoder = DenseNet121Encoder()
```

maybe throw an exception if the user passes use_7x7_first_conv_in_resnet=True